matrix_sdk/src/base_client.rs

```diff
@@ -374,14 +472,93 @@ impl Client {
     ///
     /// * `event` - The event that should be handled by the client.
     pub async fn receive_joined_state_event(&self, room_id: &RoomId, event: &StateEvent) -> bool {
-        let room_lock = self.get_or_create_room(room_id).await;
+        let room_lock = self.get_or_create_joined_room(room_id).await;
+        let mut room = room_lock.write().await;
+        room.receive_state_event(event)
+    }
+
+    /// Receive a state event for a room the user has been invited to.
+    ///
+    /// Returns true if the state of the room changed, false
+    /// otherwise.
+    ///
+    /// # Arguments
+    ///
+    /// * `room_id` - The unique id of the room the event belongs to.
+    ///
+    /// * `event` - A `AnyStrippedStateEvent` that should be handled by the client.
+    pub async fn receive_invite_state_event(
+        &self,
+        room_id: &RoomId,
+        event: &AnyStrippedStateEvent,
+    ) -> bool {
+        let room_lock = self.get_or_create_invited_room(room_id).await;
+        let mut room = room_lock.write().await;
+        room.receive_stripped_state_event(event)
+    }
+
+    /// Receive a timeline event for a room the user has left and update the client state.
+    ///
+    /// Returns a tuple of the successfully decrypted event, or None on failure and
+    /// a bool, true when the `Room` state has been updated.
+    ///
+    /// # Arguments
+    ///
+    /// * `room_id` - The unique id of the room the event belongs to.
+    ///
+    /// * `event` - The event that should be handled by the client.
+    pub async fn receive_left_timeline_event(
+        &self,
+        room_id: &RoomId,
+        event: &mut EventJson<RoomEvent>,
+    ) -> (Option<EventJson<RoomEvent>>, bool) {
+        match event.deserialize() {
+            #[allow(unused_mut)]
+            Ok(mut e) => {
+                #[cfg(feature = "encryption")]
+                let mut decrypted_event = None;
+                #[cfg(not(feature = "encryption"))]
+                let decrypted_event = None;
+
+                #[cfg(feature = "encryption")]
```

Comment:
Invited and left rooms can't receive messages and state events aren't encrypted in Matrix. This might change in the future but for now this is mostly dead code, so let's remove it.